tests/fail/tree_borrows/reservedim_spurious_write.rs

```diff
@@ -0,0 +1,109 @@
+// Illustrating a problematic interaction between Reserved, interior mutability,
+// and protectors, that makes spurious writes fail in the previous model of Tree Borrows.
+// As for all similar tests, we disable preemption so that the error message is deterministic.
+//@compile-flags: -Zmiri-tree-borrows -Zmiri-preemption-rate=0
+
+use std::cell::Cell;
+use std::sync::{Arc, Barrier};
+use std::thread;
+
+// Here is the problematic interleaving:
+// - thread 1: retag and activate `x` (protected)
+// - thread 2: create but do not retag (lazy) `y` as Reserved with interior mutability
+// - thread 1: spurious write through `x` would go here
+// - thread 2: function exit (noop due to lazyness)
+// - thread 1: function exit (no permanent effect on `y` because it is now Reserved IM unprotected)
+// - thread 2: write through `y`
+// In the source code nothing happens to `y`
+
+// `Send`able raw pointer wrapper.
+#[derive(Copy, Clone)]
+struct SendPtr(*mut u8);
+unsafe impl Send for SendPtr {}
+
+type IdxBarrier = (usize, Arc<Barrier>);
+
+// Barriers to enforce the interleaving.
+// This macro expects `synchronized!(thread, msg)` where `thread` is a `IdxBarrier`,
+// and `msg` is the message to be displayed when the thread reaches this point in the execution.
+macro_rules! synchronized {
+    ($thread:expr, $msg:expr) => {{
+        let (thread_id, barrier) = &$thread;
+        eprintln!("Thread {} executing: {}", thread_id, $msg);
+        barrier.wait();
+    }};
+}
+
+fn main() {
+    eprintln!("Without spurious write");
+    example(false);
+
+    eprintln!("\nIf this text is visible then the model forbids spurious writes.\n");
+
+    eprintln!("With spurious write");
+    example(true);
+
+    eprintln!("\nIf this text is visible then the model fails to detect a noalias violation.\n");
+}
+
+fn example(spurious: bool) {
+    // For this example it is important that we have at least two bytes
+    // because lazyness is involved.
```

Comment:
But `()` is not interior mutable. In general, using ZSTs here suggests the counterexample relies on "wonky" ZST semantics, even worse if we cook up an Interior mutable ZST type (`UnsafeCell<()>`), because what does that even mean? It would make the thing actually happening here much more obscure. And also, the counterexample is independent of whatever happens at ZST retags.

Reply:
What makes an `UnsafeCell<()>` accessed outside its range more wonky than an `UnsafeCell<u8>` outside its range?

The counterexample is independent of what happens in one of the two bytes that you have here, so having both bytes is somewhat confusing.

Reply:
Because if you have an array of `UnsafeCell<u8>` and go to the next element you naturally do an actual memory access, but if you iterate an array of `UnsafeCell<()>` you're actually doing no memory accesses at all.

Reply:
Oh I think I just understood the point. We wouldn't have two elements, we'd just have one single `Cell<u8>` that we retag as `&mut *(r as *Cell<u8> as *Cell<()>) as *mut Cell<()> as *mut Cell<u8>` I guess ?

Reply:
@JoJoDeveloping not sure why array iteration would suddenly be relevant?

@Vanille-N yes.

Reply:
I believe that @JoJoDeveloping interpreted your comment "you can have 1 byte and a `&()` reference" similarly to what I first did. Initially I thought you were arguing for using `(Cell<()>, Cell<u8>)` rather than `[Cell<u8>; 2]` but actually it would just be a single `Cell<u8>`.

Reply:
Yes I disagree with it being a single byte allocation.

Reply:
I don't understand your reasoning for that. There's a single byte that's actually relevant for this, and all this mucking about with shifting the pointer back and forth is a distraction. The key point is that the relevant byte is *not* inside the range that is initialized upon reborrow.